### DISCUSSION TITLE
ExportAction with custom query not working

### DISCUSSION THREAD

**Discussion by hristijans** [filamentphp/filament#13785] [Help]
### Package

Actions

### Package Version

v3.2.93

### How can we help you?

I have a page where I want to display some data in a table.

Here is my table definition:

```
  public function table(Table $table): Table
    {
        return $table
            ->query(OrderItem::join('products', 'order_items.product_id', '=', 'products.id')
                ->select('products.id', 'products.name', 'products.description', 'products.code', DB::raw('SUM(order_items.quantity) as quantity'))
                ->orderBy('products.id', 'desc')
                ->groupBy('products.id')
            )
            ->columns([
                TextColumn::make('code')
                    ->searchable()
                    ->label('Code'),

                TextColumn::make('name')
                    ->searchable()
                    ->label('Product Name'),

                TextColumn::make('description')
                    ->searchable()
                    ->label('Product Description'),

                TextColumn::make('quantity')
                    ->searchable()
                    ->sortable()
                    ->label('Total Quantity Sold'),
            ])
            ->headerActions([
                ExportAction::make()
                    ->exporter(OrderAllItemsExporter::class)
                    ->formats([
                        ExportFormat::Csv,
                    ])
            ])
            ->filters([
                // ...
            ])
            ->actions([
                // ...
            ])
            ->bulkActions([
                // ...
            ]);
    }
```

The table displays the results I expect and it works fine.

When i click on the Export action I am getting following notification:
![image](https://github.com/user-attachments/assets/fbf006c6-8e32-4aef-a54e-bdce3aaaaf03)

Here is my Exporter Class:

```
<?php

namespace App\Filament\Exports;

use App\Models\OrderItem;
use Filament\Actions\Exports\ExportColumn;
use Filament\Actions\Exports\Exporter;
use Filament\Actions\Exports\Models\Export;
use Illuminate\Database\Eloquent\Builder;
use Illuminate\Support\Facades\DB;

class OrderAllItemsExporter extends Exporter
{
    protected static ?string $model = OrderItem::class;

    public static function getColumns(): array
    {
        return [
            ExportColumn::make('code')
                ->label('Code'),

            ExportColumn::make('name')
                ->label('Product'),

            ExportColumn::make('quantity')
                ->label('Quantity')

        ];
    }

    public static function getCompletedNotificationBody(Export $export): string
    {
        $body = 'Your order item export has completed and '.number_format($export->successful_rows).' '.str('row')->plural($export->successful_rows).' exported.';

        if ($failedRowsCount = $export->getFailedRowsCount()) {
            $body .= ' '.number_format($failedRowsCount).' '.str('row')->plural($failedRowsCount).' failed to export.';
        }

        return $body;
    }
}

```

I am checking Telescope logs and I can notice in the `PrepareCsvExport` job there are some queries. I can see following query (I assume that is from my table query)

```
select
  distinct `products`.`id`,
  `products`.`name`,
  `products`.`description`,
  `products`.`code`,
  SUM(order_items.quantity) as quantity
from
  `order_items`
  inner join `products` on `order_items`.`product_id` = `products`.`id`
where
  `order_items`.`deleted_at` is null
group by
  `products`.`id`
order by
  `products`.`id` desc
limit
  100 offset 0
```

which returns expected data too. 

The issue is I cannot export displayed data from the table.

**Comment by eduardocm2002** [comment]
I have a similar problem, I have data in the table but the exporter doesn't show anything, maybe you solved it?

**Comment by mlselegant** [comment]
me too have same issue.

**Reply by eduardocm2002** [reply]
I'm going to try putting the query in the button, something like this example:
ExportAction::make('export') 
                    ->label('Urlaubsanträge exportieren') 
                    ->modifyQueryUsing(function ($query) { 
                        return $query->where('year', $this->year); 
                    }) 
                    ->modalHeading('Urlaubsanträge für das Jahr ' . $this->year . ' exportieren') 
                    ->formats([ExportFormat::Xlsx]) 
                    ->requiresConfirmation() 
                    ->exporter(LeaveRequestExporter::class) 
                    ->color('gray') 
                    ->icon('heroicon-o-document-arrow-down');

**Comment by CesarSarantakos** [comment]
Same problem here.
Using the default query for a table action, i catch this exception in the failed_jobs table:
`` SQL
Invalid column reference: 7 ERROR:  SELECT DISTINCT ON expressions must match initial ORDER BY expressions
``

**Reply by hristijans** [reply]
Would be nice if you post your query so we can see and help you. I believe you are missing the column in your select you are trying to order/group by.

**Reply by CesarSarantakos** [reply]
I didn't modify the Exporter query. I've only created the exporter class (based on a Model), called it on a resource (without using the modifyQueryUsing in any way) and the **default** query created by Filament is causing the error.
